### ISSUE TITLE
Missing error message when destination folder does not exist

### ISSUE THREAD

**Issue by albancho** [lovell/sharp#2999]
sharp is working well when the destination folder exists. But when the destination folder does not, the code crashes, with a generic error, which doesn't state anything about the issue.

here is the code I'm using:

```
console.log('imgToResize', imgToResize); // C:/my/path/to/origin/my-img.jpg
console.log('img960', img960); // C:/my/path/to/960/my-img.jpg (folder "960" doesn't exist)
console.log('dimensionsFull', dimensionsFull); // { height: 1335, width: 2000, type: 'jpg' }
console.log('new dim: 960x', Math.ceil(dimensionsFull.height * 960 / dimensionsFull.width)); // 641
const sharpOptions = {
    fit: 'contain',
    background: {
       r: 0,
       g: 0,
       b: 0,
       alpha: 0
    }
}
await sharp(imgToResize)
                    .resize(960, Math.ceil(dimensionsFull.height * 960 / dimensionsFull.width), sharpOptions)
                    .toFile(img960);
```

```
npm ERR! code ELIFECYCLE
npm ERR! errno 3221226505
npm ERR! my-app-name@0.0.0 imgmin: `node scripts/my-script.js`
npm ERR! Exit status 3221226505
npm ERR!
npm ERR! Failed at the my-app-name@0.0.0 my-script script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\my-user\AppData\Roaming\npm-cache\_logs\xxx-debug.log
```

The issue is solved if I add this line:
```
fse.ensureDirSync(path.dirname(img960));
```

I think sharp should check if destination folder exists at init (before sharping the actual img), and if not, throw an error which states the cause. It should be even better is sharp could make the folder if it doesn't exists, but not sure if it should.

**Comment by lovell** [comment]
I see the following error when attempting to write to a missing directory:
```
Error: does-not-exist/out.png: unable to open for write
unix error: No such file or directory
```
Perhaps you need to add some try/catch error handling to your script?

https://sharp.pixelplumbing.com/api-output#tofile

> The caller is responsible for ensuring directory structures and permissions exist.

**Comment by albancho** [comment]
Thank you for your suggestion, but I had already tried this code, with no better message:

```
            try {
                // no console.log triggered
                sharp(imgToResize)
                    .resize(960, Math.floor(dimensionsFull.height * 960 / dimensionsFull.width), sharpOptions)
                    .toFile(img960)
                    .then(res => { console.log("960 Done!", res); })
                    .catch(err => {
                        console.error("960 Error processing files", err);
                    });
            } catch(error) {
                console.log('960 error', error);
            }
```

**Comment by albancho** [comment]
> I see the following error when attempting to write to a missing directory:
> 
> ```
> Error: does-not-exist/out.png: unable to open for write
> unix error: No such file or directory
> ```

Are you on a unix environment ? I'm running it on Windows, it could be the cause...

**Comment by lovell** [comment]
I added a test for this via commit https://github.com/lovell/sharp/commit/9eada2ac5a9d55574fe85e8e958087b467fad005

https://github.com/lovell/sharp/blob/9eada2ac5a9d55574fe85e8e958087b467fad005/test/unit/io.js#L800-L811

...and it crashed on Windows:
```
npm ERR! code ELIFECYCLE
npm ERR! errno 3221226505
```
https://github.com/lovell/sharp/runs/4418122778?check_suite_focus=true

This might be a bug in libvips (or maybe glib?) - I'll investigate.

**Comment by lovell** [comment]
Possible upstream fix for this at https://github.com/libvips/libvips/pull/2571

**Comment by lovell** [comment]
v0.30.0 now available with libvips v8.12.2.
